### PR TITLE
Minion Helper

### PR DESCRIPTION
**Pull request by hannibal002**
Don't be too alarmed by the 100 changed files in 120 commits. The vast majority of changes are in the package `io/github/moulberry/notenoughupdates/miscgui/minionhelper/` and the vast majority of other classes have only a mini utils method change. 

<details>
<summary>Long / Read</summary>

## Long / Read
Adds a complex "Minion helper". The name is perhaps confusing, as the feature does not help with setting up or using minions, but with the crafting of the cheapest **unique minion upgrades** to quickly unlock the **next minion slot**. I will therefore gratefully accept ideas for a better name.

- In **the "crafted minions" inventory** (accessible via the collection page or with the command `/craftedgenerators`) a new overlay is displayed on the right side next to the inventory. On this inventory, the player sees a list of all minions he can craft, sorted by cheapest upgrade cost first.
- By default, **a filter** is active that **hides** all recipes that the player has **not yet unlocked**. This works with the **SkyBlock API**, which is queried at **every world change** to read the data from the collection tier, reputation, pelts owned and slayer level. The filter **can be switched off** with the button at the bottom right. Minions that cannot be crafted are displayed in red.
- The player can **hover** over the name of a minion and **see** the **requirements** for the minion and the exact **ingredients** of the crafting recipe listed.
- The player can **click** on the name of a minion to **open the crafting inventory** / trade inventory for the minion from NEW.
- If necessary, **several pages** are displayed, which can be scrolled through by **pressing** the **button** or the **arrow key** to the left or right.
- Shows in one line **how many more unique minion upgrades** are needed to unlock the **next minion slot**.
- **Displays pelts** the player has and needs to craft the **tier 12 farming minions**.
- Shows in the crafted minions inventory when hovering over a minion how much the upgrade costs or (via shift key togglebale) the **total cost of a minion** at a certain level.
- Supports **profile switches** and **updates** automatically when the player (or a coop member) crafts a **new unique minion**.

</details>
<details>
<summary>Images</summary>

## Images

image

image

image

image

image

</details>
<details>
<summary>Too Long / Didn’t Read</summary>

## Too Long / Didn’t Read
Adds an overlay-list showing the next cheapest unique minion upgrade, supporting requirements, pages, crafting information on hover, link to recipe inventory and amount of mission minions

### CODE REVIEW

**Comment by jani270** [comment]
Lets merge this without squashing

**Comment by NopoTheGamer** [comment]
> Lets merge this without squashing

:+1:

**Comment by efefury** [comment]
rEgEx PlEaSe

**Comment by Futuremappermydud** [comment]
image
Maybe have the page number be bold or move the buttons to the side when it needs more room?

**Comment by ThexXTURBOXx** [comment]
The only thing I have noticed is this: https://hastebin.com/aceqilerol.yaml
Otherwise, this is very nice :)

**Comment by hannibal00212** [comment]
> The only thing I have noticed is this: https://hastebin.com/aceqilerol.yaml Other than that, this is very nice :)

Thank you for testing it, really appreciate!
The errors you get are due to repo data missing. you need to manuall update to another repo version for the messages to disappear.

**Comment by ThexXTURBOXx** [comment]
That makes sense, thanks!
An additional small thing:
grafik

The tooltips don't seem to scale properly (are they maybe ignoring the gui scale?) 
For reference:
grafik

**Comment by hannibal00212** [comment]
> image Maybe have the page number be bold or move the buttons to the side when it needs more room?

Good suggestion. I will try to change the distance a bit.

**Comment by hannibal00212** [comment]
> The tooltips don't seem to scale properly (are they maybe ignoring the gui scale?)
Yes that's the case. The GUI scale getting ignored is not on purpose but I didn't get it to work properly yet. I'm open for any help on that topic.

File: src/main/java/io/github/moulberry/notenoughupdates/miscgui/minionhelper/render/MinionHelperOverlayHover.java [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+		if (mouseObject != null) {
+			GlStateManager.pushMatrix();
+			GlStateManager.scale(2f / scaledresolution.getScaleFactor(), 2f / scaledresolution.getScaleFactor(), 1);
+			Utils.drawHoveringText(getTooltip(mouseObject),
+				mouseX * scaledresolution.getScaleFactor() / 2,
+				mouseY * scaledresolution.getScaleFactor() / 2,
+				scaledWidth * scaledresolution.getScaleFactor() / 2,
+				scaledHeight * scaledresolution.getScaleFactor() / 2, -1, Minecraft.getMinecraft().fontRendererObj
+			);
+			GlStateManager.popMatrix();
```
Comment (ThexXTURBOXx) [inline src/main/java/io/github/moulberry/notenoughupdates/miscgui/minionhelper/render/MinionHelperOverlayHover.java:80]:
```suggestion
			Utils.drawHoveringText(getTooltip(mouseObject), mouseX, mouseY,
				scaledWidth, scaledHeight, -1, Minecraft.getMinecraft().fontRendererObj);
```

Comment (ThexXTURBOXx) [reply src/main/java/io/github/moulberry/notenoughupdates/miscgui/minionhelper/render/MinionHelperOverlayHover.java:80]: This should do the job

**Review by ThexXTURBOXx — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/moulberry/notenoughupdates/miscgui/minionhelper/render/MinionHelperOverlayHover.java (hunk 1 above)
Comment (hannibal00212) [reply src/main/java/io/github/moulberry/notenoughupdates/miscgui/minionhelper/render/MinionHelperOverlayHover.java]: Thanks!

**Comment by hannibal00212** [comment]
> Maybe have the page number be bold or move the buttons to the side when it needs more room?

I have rduced the distance between text and buttons. Looking better on higher page counts!
image

File: src/main/java/io/github/moulberry/notenoughupdates/core/util/StringUtils.java [hunk 2]
```diff
@@ -102,4 +102,20 @@ public static String replaceLast(String string, String toReplace, String replace
 		int start = string.lastIndexOf(toReplace);
 		return string.substring(0, start) + replacement + string.substring(start + toReplace.length());
 	}
+
+	public static String removeLastWord(String string, String splitString) {
+		try {
+			String[] split = string.split(splitString);
+			String rawTier = split[split.length - 1];
+			return string.substring(0, string.length() - rawTier.length() - 1);
+		} catch (StringIndexOutOfBoundsException e) {
```
[the remark this answers is not in the document]
Comment (romangraef) [reply src/main/java/io/github/moulberry/notenoughupdates/core/util/StringUtils.java:111]: Nvm, i forgot this language doesn't have `rsplit`

File: src/main/java/io/github/moulberry/notenoughupdates/listener/RenderListener.java [hunk 3]
```diff
@@ -1685,4 +1696,24 @@ public void onGuiScreenKeyboard(GuiScreenEvent.KeyboardInputEvent.Pre event) {
 	public void onRenderLast(RenderWorldLastEvent event) {
 		CrystalMetalDetectorSolver.render(event.partialTicks);
 	}
+
+	/**
+	 * Support for switching between different pages in the RecipeView gui via right and left arrow key
+	 * @param event
+	 */
+	//Because GuiScreen.keyTyped does not fire the KEY_LEFT and KEY_RIGHT keys. Maybe some event cancelled it?
+	@SubscribeEvent
```
Comment (romangraef) [inline src/main/java/io/github/moulberry/notenoughupdates/listener/RenderListener.java:1705]: i love making a 2000 lines file even longer with code relating to a feature that is entirely coded in a completely different class

**Comment by IRONM00N** [comment]
this is cool

**Comment by hannibal002** [comment]
Wait for #351 to merge first, so I can merge developer list again

**Comment by hannibal002** [comment]
This PR is finally ready to merge, at least from my side. It's ready to review ;)

Sorry, nea. I've forgot about your extra effort in splitting it into two prs. I did already new changes, so it's too late for using these :/

File: src/main/java/io/github/moulberry/notenoughupdates/listener/RenderListener.java (hunk 3 above)
Comment (hannibal002) [reply src/main/java/io/github/moulberry/notenoughupdates/listener/RenderListener.java:1735]: Is this comment still a valid change request? Should I find another class to put this method in? Or just create a new class entirely?

**Review by romangraef — APPROVED** [review]

**Review by Lulonaut — APPROVED** [review]

**Review by NopoTheGamer — APPROVED** [review]

**Review by romangraef — APPROVED** [review]
2.1.1.md

**Pull request merged**